getCountQuery() fails on entity-view with self-reference sub

`PaginatedCriteriaBuilder.getCountQuery()` throws `org.hibernate.query.SemanticException: Could not interpret path expression 'reversesMovement_1'` when:

1. The projected `EntityView` declares a **self-reference subview** (same target entity, different view type — required to avoid the recursive-subview cycle Blaze rejects at boot).
2. The underlying `CriteriaBuilder` filters on the **same association** by name, e.g. `cb.where("m.reversesMovement").isNull()`.

The fetch-join Blaze emits for the subview gets a Hibernate-generated alias (`reversesMovement_1`). The COUNT query the paginated builder emits later resolves the WHERE-clause path against that alias and fails.

Walking the FK column directly (`cb.where("m.reversesMovement.id").isNull()`) sidesteps the bug.

### Versions

- `blaze-persistence` **1.6.18** (`-core-impl`, `-entity-view-impl`,
  `-integration-hibernate-7.2`)
- `hibernate-orm` **7.2.12.Final**
- Jakarta Persistence 3.2
- PostgreSQL 17 (Testcontainers)
- JDK 25

### Minimal reproducer

https://github.com/heruan/blaze-issues/pull/2 (branch `repro/paginated_count_self_subview`)

```sh
./gradlew test --tests 'to.lova.blaze.issues.paginated_count_self_subview.*'
```

### Entity

```java
@Entity
public class Movement {
    @Id @GeneratedValue Long id;
    @Column(nullable = false) LocalDate date;
    @Column(nullable = false) BigDecimal amount;
    @ManyToOne Movement reversesMovement;
}
```

### Entity views

```java
@EntityView(Movement.class)
public interface MovementView {
    @IdMapping Long getId();
    LocalDate getDate();
    BigDecimal getAmount();
    MovementReferenceView getReversesMovement();   // self-reference subview
}

@EntityView(Movement.class)
public interface MovementReferenceView {
    @IdMapping Long getId();
    LocalDate getDate();
    BigDecimal getAmount();
}
```

### Failing call

```java
EntityViewSetting<MovementView, ?> setting = EntityViewSetting.create(MovementView.class, 0, 50);
setting.addAttributeSorter("id", Sorters.ascending());

var cb = cbf.create(em, Movement.class, "m");
cb.where("m.reversesMovement").isNull();           // <-- direct association

evm.applySetting(setting, cb).getResultList();     // boom on COUNT
```

### Expected

The paginated query runs and returns the seeded rows (here: 2, both with `reversesMovement` null).

### Actual

```
java.lang.IllegalArgumentException: org.hibernate.query.SemanticException:
        Could not interpret path expression 'reversesMovement_1'
    at org.hibernate.internal.ExceptionConverterImpl.convert(...)
    at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:1010)
    at org.hibernate.internal.SessionImpl.createQuery(SessionImpl.java:155)
    at com.blazebit.persistence.impl.PaginatedCriteriaBuilderImpl.getCountQuery(PaginatedCriteriaBuilderImpl.java:452)
    at com.blazebit.persistence.impl.PaginatedCriteriaBuilderImpl.getQuery(PaginatedCriteriaBuilderImpl.java:528)
    at com.blazebit.persistence.impl.PaginatedCriteriaBuilderImpl.getResultList(PaginatedCriteriaBuilderImpl.java:604)
    ...
Caused by: org.hibernate.query.SemanticException:
        Could not interpret path expression 'reversesMovement_1'
    at org.hibernate.query.hql.internal.BasicDotIdentifierConsumer$BaseLocalSequencePart.resolveTerminal(BasicDotIdentifierConsumer.java:180)
    at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitSimplePath(SemanticQueryBuilder.java:5786)
    at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitUnaryIsPredicate(SemanticQueryBuilder.java:2407)
    at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitWhereClause(SemanticQueryBuilder.java:2336)
    ...
    at org.hibernate.query.hql.internal.StandardHqlTranslator.translate(StandardHqlTranslator.java:64)
```

The crash is on the COUNT half of the paginated pair; the data query alone (without `EntityViewSetting`-driven pagination) does not reach the failing code path.

### Workaround

Walk the FK column directly via `.id` instead of the association name. This bypasses the implicit-join codepath that collapses onto the fetch alias and the count emitter is happy:

```java
cb.where("m.reversesMovement.id").isNull();   // ✅ passes on 1.6.18
```

Both forms are covered by the reproducer (`crashesOnDirectAssociationIsNull` and `workaroundFkPathSucceeds`) so the workaround acts as a regression guard for the day the upstream fix lands and the idiomatic association form starts working again.

### Related observation in production code

In a real GraphQL projection (large fetch tree, multiple filters), the correlated subquery form

```java
cb.whereNotExists()
    .from(Movement.class, "rev")
    .where("rev.reversesMovement.id").eqExpression("m.id")
    .end();
```

crashes with the same `Could not interpret path expression 'reversesMovement_1'`, and `OUTER(id)` does not help — Hibernate resolves the path against the fetch alias before the `OUTER` substitution runs. We fall back to an explicit `LEFT JOIN ... ON ... WHERE alias.id IS NULL` anti-join, at the cost of Postgres no longer recognising the predicate as a semantic anti-join in `EXPLAIN`. The minimal repro here only triggers the direct-association variant; the correlated NOT EXISTS variant did not reproduce in the trimmed setup, so an additional factor (fetch tree depth?) appears to be required.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ None yet
 ### Bug fixes
 
 * Fix `StringIndexOutOfBoundsException` when using a `VALUES` clause and an inline CTE join inside another inline CTE
+* Fix count and id query omitting joins that the WHERE clause refers to through an association path that is implicitly joined in the SELECT clause
 
 ### Backwards-incompatible changes
 

--- a/core/impl/src/main/java/com/blazebit/persistence/impl/AbstractFullQueryBuilder.java
+++ b/core/impl/src/main/java/com/blazebit/persistence/impl/AbstractFullQueryBuilder.java
@@ -356,7 +356,7 @@ public abstract class AbstractFullQueryBuilder<T, X extends FullQueryBuilder<T, 
             } else {
                 // Collect usage of collection join nodes to optimize away the count distinct
                 // Note that we always exclude the nodes with group by dependency. We consider just the ones from the identifiers
-                Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getIdentifierExpressionsToUseNonRootJoinNodes();
+                Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getCountAndIdQueryAlwaysIncludedNodes(getIdentifierExpressionsToUseNonRootJoinNodes());
                 Set<JoinNode> collectionJoinNodes = joinManager.buildClause(sbSelectFrom, clauseExclusions, null, true, externalRepresentation, true, false, optionalWhereClauseConjuncts, whereClauseConjuncts, explicitVersionEntities, countNodesToFetch, identifierExpressionsToUseNonRootJoinNodes, null, true);
                 boolean hasCollectionJoinUsages = collectionJoinNodes.size() > 0;
 
@@ -459,7 +459,7 @@ public abstract class AbstractFullQueryBuilder<T, X extends FullQueryBuilder<T, 
             } else {
                 // Collect usage of collection join nodes to optimize away the count distinct
                 // Note that we always exclude the nodes with group by dependency. We consider just the ones from the identifiers
-                Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getIdentifierExpressionsToUseNonRootJoinNodes();
+                Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getCountAndIdQueryAlwaysIncludedNodes(getIdentifierExpressionsToUseNonRootJoinNodes());
                 Set<JoinNode> collectionJoinNodes = joinManager.buildClause(sbSelectFrom, COUNT_QUERY_GROUP_BY_CLAUSE_EXCLUSIONS, null, true, externalRepresentation, true, false, optionalWhereClauseConjuncts, whereClauseConjuncts, explicitVersionEntities, countNodesToFetch, identifierExpressionsToUseNonRootJoinNodes, null, true);
                 boolean hasCollectionJoinUsages = collectionJoinNodes.size() > 0;
 
@@ -603,6 +603,33 @@ public abstract class AbstractFullQueryBuilder<T, X extends FullQueryBuilder<T, 
         }
 
         return joinNodes;
+    }
+
+    protected Set<JoinNode> getCountAndIdQueryAlwaysIncludedNodes(Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes) {
+        // Path expressions in the WHERE clause might lazily resolve to default join nodes that were created
+        // for other clauses only e.g. an implicit join in the SELECT clause. Since such nodes don't have a
+        // WHERE clause dependency registered, they would be omitted from count and id queries even though
+        // the rendered WHERE clause refers to their alias, so we have to include them explicitly
+        List<JoinNode> whereClauseJoinNodes = new ArrayList<>();
+        whereManager.acceptVisitor(new JoinNodeGathererVisitor(whereClauseJoinNodes));
+        Set<JoinNode> alwaysIncludedNodes = null;
+        for (int i = 0; i < whereClauseJoinNodes.size(); i++) {
+            JoinNode node = whereClauseJoinNodes.get(i);
+            while (node != null && node.getParent() != null) {
+                if (alwaysIncludedNodes == null) {
+                    if (identifierExpressionsToUseNonRootJoinNodes.contains(node)) {
+                        break;
+                    }
+                    alwaysIncludedNodes = new HashSet<>(identifierExpressionsToUseNonRootJoinNodes);
+                }
+                if (!alwaysIncludedNodes.add(node)) {
+                    break;
+                }
+                alwaysIncludedNodes.addAll(node.getDependencies());
+                node = node.getParent();
+            }
+        }
+        return alwaysIncludedNodes == null ? identifierExpressionsToUseNonRootJoinNodes : alwaysIncludedNodes;
     }
 
     protected ResolvedExpression[] getIdentifierExpressions() {

--- a/core/impl/src/main/java/com/blazebit/persistence/impl/PaginatedCriteriaBuilderImpl.java
+++ b/core/impl/src/main/java/com/blazebit/persistence/impl/PaginatedCriteriaBuilderImpl.java
@@ -1209,7 +1209,7 @@ public class PaginatedCriteriaBuilderImpl<T> extends AbstractFullQueryBuilder<T,
         // The id query does not have any fetch owners
         // Note that we always exclude the nodes with group by dependency. We consider just the ones from the identifiers
         Set<JoinNode> idNodesToFetch = Collections.emptySet();
-        Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getIdentifierExpressionsToUseNonRootJoinNodes();
+        Set<JoinNode> identifierExpressionsToUseNonRootJoinNodes = getCountAndIdQueryAlwaysIncludedNodes(getIdentifierExpressionsToUseNonRootJoinNodes());
         Set<JoinNode> collectionJoins = joinManager.buildClause(sbSelectFrom, ID_QUERY_GROUP_BY_CLAUSE_EXCLUSIONS, null, true, externalRepresentation, true, false, optionalWhereClauseConjuncts, whereClauseConjuncts, explicitVersionEntities, idNodesToFetch, identifierExpressionsToUseNonRootJoinNodes, null, true);
 
         if (keysetMode == KeysetMode.NONE || keysetManager.getKeysetLink().getKeyset().getTuple() == null) {

--- a/core/testsuite/src/test/java/com/blazebit/persistence/testsuite/PaginationTest.java
+++ b/core/testsuite/src/test/java/com/blazebit/persistence/testsuite/PaginationTest.java
@@ -241,6 +241,24 @@ public class PaginationTest extends AbstractCoreTest {
         assertEquals("doc1", result.get(0).getName());
     }
 
+    // Test for #2114
+    @Test
+    public void testPaginatedWhereOnAssociationImplicitlyJoinedInSelect() {
+        String expectedCountQuery = "SELECT " + countPaginated("d.id", false) + " FROM Document d LEFT JOIN d.parent parent_1 WHERE parent_1 IS NULL";
+        String expectedObjectQuery = "SELECT d.id, parent_1.name FROM Document d LEFT JOIN d.parent parent_1 WHERE parent_1 IS NULL ORDER BY d.id ASC";
+        PaginatedCriteriaBuilder<Tuple> cb = cbf.create(em, Tuple.class)
+                .from(Document.class, "d")
+                .where("d.parent").isNull()
+                .select("d.id")
+                .select("d.parent.name")
+                .orderByAsc("d.id")
+                .page(0, 10);
+        assertEquals(expectedCountQuery, cb.getPageCountQueryString());
+        assertEquals(expectedObjectQuery, cb.withInlineCountQuery(false).getQueryString());
+        PagedList<Tuple> result = cb.getResultList();
+        assertEquals(7, result.getTotalSize());
+    }
+
     @Test
     public void testSelectIndexedWithParameter() {
         String expectedCountQuery = "SELECT " + countPaginated("d.id", false) + " FROM Document d JOIN d.owner owner_1 WHERE owner_1.name = :param_0";

